pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		log.Warnf("failed to load /etc/resolv.conf: %v", err)
+		return nil, err
+	}
+
+	// Unlike traditional DNS resolvers, we do not need to append the search
+	// namespace to a given query and try to resolve it. This is because the
+	// agent acts as a DNS interceptor for DNS queries made by the application.
+	// The application's resolver is already sending us DNS queries, one for each
+	// of the DNS search namespaces. We simply need to check the existence of this
+	// name in our local nametable. If not, we will forward the query to the
+	// upstream resolvers as is.
+	//
+	// NOTE: There is still a NxN issue here as the app will cycle through N
+	// nameservers for each potential FQDN string. The agent will also cycle through
+	// NxN unfortunately as it cannot infer the orig dst IP from the incoming packet.
+	if dnsConfig != nil {
+		if len(dnsConfig.Servers) > 0 {
+			for _, s := range dnsConfig.Servers {
+				h.resolvConfServers = append(h.resolvConfServers, s+":53")
+			}
+		}
+		h.searchNamespaces = dnsConfig.Search
+	}
+	return h, nil
+}
+
+// StartDNS starts the DNS-over-UDP downstreamServer.
+func (h *LocalDNSServer) StartDNS() {
+	var err error
+	h.downstreamServer = &dns.Server{Handler: h.downstreamMux}
+	h.downstreamServer.PacketConn, err = net.ListenPacket("udp", ":15053")
+	if err != nil {
+		log.Warna("Failed to start DNS downstreamServer", err)
+	}
+
+	log.Infoa("Started DNS downstreamServer at 0.0.0.0:15053")
+	go func() {
+		err := h.downstreamServer.ActivateAndServe()
+		if err != nil {
+			log.Errora("Failed to activate DNS downstreamServer ", err)
+		}
+	}()
+}
+
+func (h *LocalDNSServer) UpdateLookupTable(nt *nds.NameTable) {
+	lookupTable := &LookupTable{
+		name4: map[string][]dns.RR{},
+		name6: map[string][]dns.RR{},
+		cname: map[string][]dns.RR{},
+	}
+	for host, ni := range nt.Table {
+		// Given a host
+		// if its a non-k8s host, store the host+. as the key with the pre-computed DNS RR records
+		// if its a k8s host, store all variants (i.e. shortname+., shortname+namespace+., fqdn+., etc.)
+		// shortname+. is only for hosts in current namespace
+		var altHosts []string
+		if ni.Registry == "Kubernetes" {
+			altHosts = generateAltHosts(host, ni, h.proxyNamespace, h.proxyDomain, h.proxyDomainParts)
+		} else {
+			altHosts = []string{host + "."}
+		}
+		ipv4, ipv6 := separateIPtypes(ni.Ips)
+		if len(ipv6) == 0 && len(ipv4) == 0 {
+			// malformed ips
+			continue
+		}
+		lookupTable.buildDNSAnswers(altHosts, ipv4, ipv6, h.searchNamespaces)
+	}
+	h.lookupTable.Store(lookupTable)
+}
+
+// ServerDNS is the implementation of DNS interface
+func (h *LocalDNSServer) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
+	var response *dns.Msg
+
+	if len(req.Question) == 0 {
+		response = new(dns.Msg)
+		response.SetReply(req)
+		response.Rcode = dns.RcodeNameError
+	} else {
+		// we expect only one question in the query even though the spec allows many
```

Comment:
is this a safe assumption? I don't know enough about DNS to answer that but it seems suspicious

Reply:
Yep. This is what the entire planet does :). Including coredns - their entire code hinges on this assumption.